subsys/bluetooth/host/gatt.c

```diff
@@ -180,13 +183,122 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
 #endif
 );
 
-#if defined(CONFIG_BT_GATT_DYNAMIC_DB)
-static void sc_ccc_cfg_changed(const struct bt_gatt_attr *attr,
-			       u16_t value)
+#if defined(CONFIG_BT_GATT_SERVICE_CHANGED)
+struct sc_data {
+	u16_t start;
+	u16_t end;
+} __packed;
+
+struct gatt_sc_cfg {
+	u8_t		id;
+	bt_addr_le_t	peer;
+	struct sc_data	data;
+};
+
+#define SC_CFG_MAX (CONFIG_BT_MAX_PAIRED + CONFIG_BT_MAX_CONN)
+static struct gatt_sc_cfg sc_cfg[SC_CFG_MAX] = {};
+
+static struct gatt_sc_cfg *find_sc_cfg(u8_t id, bt_addr_le_t *addr)
+{
+	BT_DBG("id: %u, addr: %s", id, bt_addr_le_str(addr));
+
+	for (size_t i = 0; i < ARRAY_SIZE(sc_cfg); i++) {
+		if (id == sc_cfg[i].id &&
+		    !bt_addr_le_cmp(&sc_cfg[i].peer, addr)) {
+			return &sc_cfg[i];
+		}
+	}
+
+	return NULL;
+}
+
+#if defined(CONFIG_BT_SETTINGS)
+static int sc_store(struct gatt_sc_cfg *cfg)
+{
+	char key[BT_SETTINGS_KEY_MAX];
+	int err;
+
+	if (cfg->id) {
+		char id_str[4];
+
+		u8_to_dec(id_str, sizeof(id_str), cfg->id);
+		bt_settings_encode_key(key, sizeof(key), "sc",
+				       &cfg->peer, id_str);
+	} else {
+		bt_settings_encode_key(key, sizeof(key), "sc",
+				       &cfg->peer, NULL);
+	}
+
+	if (!cfg->data.start && !cfg->data.end) {
+		err = settings_delete(key);
+	} else {
+		err = settings_save_one(key, (char *)&cfg->data,
+					sizeof(cfg->data));
+	}
+
+	if (err) {
+		BT_ERR("failed to store SC (err %d)", err);
+		return err;
+	}
+
+	BT_DBG("stored SC for %s (%s, %u-%u)", bt_addr_le_str(&cfg->peer),
+	       log_strdup(key), cfg->data.start, cfg->data.end);
+
+	return 0;
+}
+#endif /* CONFIG_BT_SETTINGS */
+
+static void sc_clear(struct gatt_sc_cfg *cfg)
+{
+#if defined(CONFIG_BT_SETTINGS)
+	bool modified = false;
+
+	if (cfg->data.start || cfg->data.end) {
+		modified = true;
+	}
+#endif /* CONFIG_BT_SETTINGS */
+
+	BT_DBG("peer %s", bt_addr_le_str(&cfg->peer));
+
+	cfg->data.start = 0;
+	cfg->data.end = 0;
+
+#if defined(CONFIG_BT_SETTINGS)
+	if (modified && bt_addr_le_is_bonded(cfg->id, &cfg->peer)) {
+		sc_store(cfg);
+	}
+#endif /* CONFIG_BT_SETTINGS */
+
+	cfg->id = BT_ID_DEFAULT;
```

Comment:
Do you mean I should use 0 instead of BT_ID_DEFAULT for this assignment (it's a u8_t so memset might be overkill)? The next line already uses memset (bt_addr_le_copy) to set the 'peer'.

Reply:
No I meant. memset(cfg, 0, sizeof(*cfg));

Reply:
Just a suggestion to memset the cfg struct. As all the assignments are to 0.

Reply:
Yes of course, I agree it makes it simpler... Done in latest push!